connectivity/service/src/main/java/org/eclipse/ditto/connectivity/service/mapping/CloudEventsMapper.java

```diff
@@ -0,0 +1,187 @@
+    final JsonObject jsonObject = newObject(payload);
+    if (jsonObject.getValue(SPECVERSION).isPresent() && jsonObject.getValue(TYPE).isPresent()
+        && jsonObject.getValue(ID).isPresent() && jsonObject.getValue(SOURCE).isPresent()) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  private JsonifiableAdaptable extractData(final String message) throws Exception {
+    final JsonObject payloadJson = newObject(message);
+    if (payloadJson.getValue("data_base64").isPresent()) {
+      final String base64Data = payloadJson.getValue("data_base64").get().asString();
+      final String decodedData = base64decoding(base64Data);
+      final JsonifiableAdaptable decodedJsonifiableAdaptable = DittoJsonException.wrapJsonRuntimeException(
+          () -> ProtocolFactory.jsonifiableAdaptableFromJson(newObject(decodedData)));
+      return decodedJsonifiableAdaptable;
+    }
+    if (payloadJson.getValue("data").isPresent()) {
+      final String data = payloadJson.getValue("data").get().toString();
+      return DittoJsonException.wrapJsonRuntimeException(
+          () -> ProtocolFactory.jsonifiableAdaptableFromJson(newObject(data)));
+    } else {
+      throw new Exception("Invalid CloudEvent");
+    }
+  }
+
+  public boolean checkHeaders(final ExternalMessage message) {
+    Map<String, String> headers = message.getHeaders();
+    if (headers.get(CE_ID) == null || headers.get(CE_SOURCE) == null
+        || headers.get(CE_TYPE) == null || headers.get(CE_SPECVERSION) == null) {
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+
+  @Override
+  public List<Adaptable> map(final ExternalMessage message) {
+    // build Exception
+    MessageMappingFailedException mappingFailedException = MessageMappingFailedException.newBuilder(
+            message.findContentType().orElse("")).description("This is not a CloudEvent")
+        .dittoHeaders(DittoHeaders.of(message.getHeaders())).build();
+    if (message.findContentType().get().equals(contentType)) {
+      // extract message as String
+      final String payload = extractPayloadAsString(message);
+      try {
+        if (checkHeaders(message)) {
+          final JsonifiableAdaptable binaryAdaptable = DittoJsonException.wrapJsonRuntimeException(
+              () -> ProtocolFactory.jsonifiableAdaptableFromJson(newObject(payload)));
+          DittoHeaders headers = binaryAdaptable.getDittoHeaders();
+          return singletonList(
+              ProtocolFactory.newAdaptableBuilder(binaryAdaptable).withHeaders(headers).build());
+        }
+        if (validatePayload(payload)) {
+          final JsonifiableAdaptable adaptable = extractData(payload);
+          final DittoHeaders headers = adaptable.getDittoHeaders();
+          return singletonList(
+              ProtocolFactory.newAdaptableBuilder(adaptable).withHeaders(headers).build());
+        } else {
+          throw mappingFailedException;
+        }
+      } catch (Exception e) {
```

Comment:
You should not just catch *any*  exception and replace it with another - that way the root "cause" will be lost.
Build and throw a new MessageMappingFailedException with the catched exception as `cause`.